api/krusty/repeatbase_test.go

```diff
@@ -128,7 +128,8 @@ spec:
 
 	err := th.RunWithErr("mango", th.MakeDefaultOptions())
 	if !strings.Contains(
-		err.Error(), "multiple matches for Id apps_v1_Deployment|~X|banana; failed to find unique target for patch") {
+		err.Error(), "multiple matches for Id {group: apps, version: v1, kind: Deployment, name: banana};"+
+			" failed to find unique target for patch {group: apps, version: v1, kind: Deployment, name: banana}") {
```

Comment:
> This error message highlights a hesitation I have about the proposed format: it looks like an object, not a singular identifier, so in a context like this, the error message reads as though the the identifier is the patch. Plus, the format is now very close to the full object/type meta expression, yet does not match it exactly.

This is a fair point, and I agree that it would be good to main some consistency with what other projects do to stringify GVKNN. 

I like your proposal of `Kind.group.version/name.namespace` - especially since it aligns with some kubectl parsing; however I would suggest we do `Kind.version.group/name.namespace` since that is what I am seeing more frequently in the links you've provided. Another thing is that with this format,  it is less clear to me what to do in the scenario that something is missing. Leaving out the namespace when there is none (e.g. `ClusterRole.v1.rbac.authorization.k8s.io/my-cluster-role`) is harmless enough, but what about other cases?

For a missing name, I'm inclined to think that if we are missing a name, we will also be missing a namespace (I'm struggling to think of a valid example of a ResId that has a namespace but no name), in which case IMO we can just print out the gvk - `Kind.version.group`. 

If, however, one or more of `Kind`, `version`, or `group` is missing - which unless I am mistaken can sometimes happen with patches or replacement sources or other local configuration files - I think we will need a placeholder. The existing placeholders `~G`,`~V`, etc, though unambigious and perhaps helpful to us as maintainers who are intimately familiar with kustomize, are unintuitive IMO. Seeing something like `ClusterRole.~V.~G/my-cluster-role` is also quite difficult to read and understand at first glance. 

I wonder if for a missing group, kind, or version, we can explicitly have empty brackets to signify that something is empty, e.g. for a missing group and version, something like `ClusterRole.{}.{}/my-cluster-role` This is still succinct and unambiguous and IMO a bit easier to read than something like `ClusterRole.~V.~G/my-cluster-role`. Some alternative ideas:

- `ClusterRole.{}.{}/my-cluster-role`
- `ClusterRole.[].[]/my-cluster-role`
- `ClusterRole."".""/my-cluster-role`
- `ClusterRole.[no version].[no group]/my-cluster-role`

WDYT?

Reply:
> however I would suggest we do Kind.version.group/name.namespace since that is what I am seeing more frequently in the links you've provided.

🤦‍♀️ good catch

> If, however, one or more of Kind, version, or group is missing - which unless I am mistaken can sometimes happen with patches or replacement sources or other local configuration files - I think we will need a placeholder.

I agree, and am leaning towards always using the placeholder rather than only in some circumstances.

The first two examples remind me of JSONpath notation (and our related field spec notation), which I'm not sure is a good thing. The third is an intriguing option... my aesthetic reaction is that it looks weird lol, but it is readable, and the logic that we quote an empty string to indicate the value is missing is intuitive enough. I was leaning towards the fourth option, but then I found this interesting bit of historical context: https://github.com/kubernetes-sigs/kustomize/pull/620. I'm noticing there that the "before" placeholder values make the real values less apparent at a glance. The brackets in `ClusterRole.[no version].[no group]/my-cluster-role` (or maybe `ClusterRole.[noVersion].[noGroup]/my-cluster-role.[noNamespace]`) help, but the real values don't stand out as much as in the quotes option. 

Playing with the "this is blank" visual from your quotes idea:
`ClusterRole._._/my-cluster-role._`
`ClusterRole.-.-/my-cluster-role.-`
With all of these variants, I am a little worried about the ambiguity that remains (for users, not parsing) by relying on ordering. If either version or group are present, it's probably obvious which it is. But that's not necessarily true for name and namespace, or when most everything is blank.

Maybe the brackets idea, but trying to shorten the names?
`ClusterRole.[noVer].[noGrp]/my-cluster-role.[noNs]`
fully blank: `[noKind].[noVer].[noGrp]/[noName].[noNs]`

Reply:
> With all of these variants, I am a little worried about the ambiguity that remains (for users, not parsing) by relying on ordering. If either version or group are present, it's probably obvious which it is. But that's not necessarily true for name and namespace, or when most everything is blank.

That's fair, something like ""."".""/""."" or _._._/_._ would not be very helpful or intuitive. I'm happy with using brackets and shorter names: `ClusterRole.[noVer].[noGrp]/my-cluster-role.[noNs]`, `[noKind].[noVer].[noGrp]/[noName].[noNs]`. I will update the PR shortly.

Unimportant tidbit, after taking another look at #620, I've finally realized that the tilde has meaning - the bitwise/logical symbol for "not"! I feel a bit silly for never having understood it before but it makes sense to me now 😂

Reply:
Updated!